presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/ApplyTableScanRedirect.java

```diff
@@ -0,0 +1,123 @@
+import static java.lang.String.format;
+import static java.util.Objects.requireNonNull;
+
+public class ApplyTableScanRedirect
+        implements Rule<TableScanNode>
+{
+    private static final Pattern<TableScanNode> PATTERN = tableScan();
+
+    private final Metadata metadata;
+    private final DomainTranslator domainTranslator;
+
+    public ApplyTableScanRedirect(Metadata metadata)
+    {
+        this.metadata = requireNonNull(metadata, "metadata is null");
+        this.domainTranslator = new DomainTranslator(metadata);
+    }
+
+    @Override
+    public Pattern<TableScanNode> getPattern()
+    {
+        return PATTERN;
+    }
+
+    @Override
+    public boolean isEnabled(Session session)
+    {
+        return isTableScanRedirectionEnabled(session);
+    }
+
+    @Override
+    public Result apply(TableScanNode node, Captures captures, Context context)
+    {
+        Optional<TableScanRedirectApplicationResult> tableScanRedirectApplicationResult = metadata.applyTableScanRedirect(context.getSession(), node.getTable());
+        if (tableScanRedirectApplicationResult.isEmpty()) {
+            return Result.empty();
+        }
+
+        CatalogSchemaTableName destinationTable = tableScanRedirectApplicationResult.get().getDestinationTable();
+        Optional<TableHandle> destinationTableHandle = metadata.getTableHandle(
+                context.getSession(),
+                convertFromSchemaTableName(destinationTable.getCatalogName()).apply(destinationTable.getSchemaTableName()));
+        if (destinationTableHandle.isEmpty()) {
+            throw new PrestoException(GENERIC_INTERNAL_ERROR, format("Destination table %s from table scan redirection not found", destinationTable));
+        }
+        if (destinationTableHandle.get().equals(node.getTable())) {
+            return Result.empty();
+        }
+        Map<ColumnHandle, String> columnMapping = tableScanRedirectApplicationResult.get().getDestinationColumns();
+        Map<String, ColumnHandle> destinationColumnHandles = metadata.getColumnHandles(context.getSession(), destinationTableHandle.get());
+        Map<Symbol, ColumnHandle> newAssignments = node.getAssignments().entrySet().stream()
+                .collect(toImmutableMap(Map.Entry::getKey, entry -> {
+                    String destinationColumn = requireNonNull(
+                            columnMapping.get(entry.getValue()),
+                            format("Did not find mapping for source column %s in table scan redirection", entry.getValue()));
+                    return requireNonNull(
+                            destinationColumnHandles.get(destinationColumn),
+                            format("Did not find handle for column %s in destination table %s", destinationColumn, destinationTable));
+                }));
+
+        TableScanNode newScanNode = new TableScanNode(
+                node.getId(),
+                destinationTableHandle.get(),
+                node.getOutputSymbols(),
```

Comment:
this will prune columns from dest TS, but there might be predicate on non-mapped columns, e.g:
```
mapping:
src_handle -> "dest_col1"
filter:
TupleDomain "dest_col2" IN (1, 2)
```
currently, TS will only output `dest_col1`, so new filter on `dest_col2` will have missing symbols

Example query that should shows this:
```
SELECT a FROM table where b > 0
```
In this case `b > 0` might be fully consumed by connector and `b` column might be pruned in TS

Reply:
Fixed now